pyscf/pbc/dft/multigrid/multigrid_pair.py

```diff
@@ -655,79 +596,48 @@ def nr_rks(mydf, xc_code, dm_kpts, hermi=1, kpts=None,
     coulG = tools.get_coulG(cell, mesh=mesh)
 
     #:vG = np.einsum('ng,g->ng', rhoG[:,0], coulG)
-    vG = np.empty_like(rhoG[:,0], dtype=np.result_type(rhoG[:,0], coulG))
-    for i, rhoG_i in enumerate(rhoG[:,0]):
-        vG[i] = np.multiply(rhoG_i, coulG, out=vG[i])
+    vG = np.multiply(rhoG[0,0], coulG)
     coulG = None
 
     if mydf.vpplocG_part1 is not None:
-        for i in range(nset):
-            vG[i] += mydf.vpplocG_part1 * 2
+        vG += mydf.vpplocG_part1 * 2
 
     #:ecoul = .5 * np.einsum('ng,ng->n', rhoG[:,0].real, vG.real)
     #:ecoul+= .5 * np.einsum('ng,ng->n', rhoG[:,0].imag, vG.imag)
-    ecoul = np.zeros((rhoG.shape[0],))
-    for i in range(rhoG.shape[0]):
-        ecoul[i] = .5 * np.vdot(rhoG[i,0], vG[i]).real
+    ecoul = .5 * np.vdot(rhoG[0,0], vG).real
 
     ecoul /= cell.vol
     log.debug('Multigrid Coulomb energy %s', ecoul)
 
     if mydf.vpplocG_part1 is not None:
-        for i in range(nset):
-            vG[i] -= mydf.vpplocG_part1
+        vG -= mydf.vpplocG_part1
 
     weight = cell.vol / ngrids
     # *(1./weight) because rhoR is scaled by weight in _eval_rhoG.  When
     # computing rhoR with IFFT, the weight factor is not needed.
     rhoR = tools.ifft(rhoG.reshape(-1,ngrids), mesh).real * (1./weight)
-    rhoR = rhoR.reshape(nset,-1,ngrids)
-    wv_freq = []
-    nelec = np.zeros(nset)
-    excsum = np.zeros(nset)
-    for i in range(nset):
-        exc, vxc = ni.eval_xc(xc_code, rhoR[i], spin=0, deriv=1)[:2]
-        if xctype == 'LDA':
-            wv = weight * vxc[0]
-            wv_freq.append(tools.fft(wv, mesh))
-            wv = None
-        elif xctype == 'GGA':
-            if GGA_METHOD.upper() == 'FFT':
-                wv_freq.append(_rks_gga_wv0_pw(cell, rhoR[i], vxc, weight, mesh).reshape(1,ngrids))
-            else:
-                wv = _rks_gga_wv0(rhoR[i], vxc, weight)
-                wv_freq.append(tools.fft(wv, mesh))
-                wv = None
-        elif vxc is None:
-            wv_freq.append(np.zeros((1,*mesh), dtype=np.complex128))
-        else:
-            raise NotImplementedError
-
-        nelec[i]  += np.sum(rhoR[i,0]) * weight
-        excsum[i] += np.sum(rhoR[i,0] * exc) * weight
-        exc = vxc = None
-
+    rhoR = rhoR.reshape(-1,ngrids)
+    exc, vxc = ni.eval_xc_eff(xc_code, rhoR, deriv=1, xctype=xctype)[:2]
+    wv = weight * vxc
+    wv_freq = tools.fft(wv, mesh).reshape(-1,ngrids)
+    if xctype == 'GGA' and GGA_METHOD.upper() == 'FFT':
+        Gv = cell.get_Gv(ni.mesh)
+        wv_freq[0] -= np.einsum('xp,px->p', 1j*wv_freq[1:4], Gv)
```

Comment:
This einsum was slow for large grids, and that's why I added `get_gga_vrho_gs`, which is a multithreaded version. I'm not sure whether numpy 2.3 has better performance for this, but it is worth checking. We should check the wall time for `get_veff` with multithreading.